### PR TITLE
Multiselect files in dataset + download folders

### PR DESCRIPTION
**Pull request by max-zilla**
Screen Shot 2020-06-23 at 10 53 25 AM

Enable 'marking' of files inside a dataset for bulk actions:
- download all, tag all, delete all

Marked files are stored in a cookie for each dataset and saved if you leave the page. Currently downloading feature for both file selections and folders uses the streaming code w/ bagit and other metadata, if we want a simpler package that will need to be changed.

### CODE REVIEW

File: app/views/datasets/filesAndFolders.scala.html [hunk 1]
```diff
@@ -27,47 +31,292 @@
...
+            } else {
+                notify("No files currently marked", "warning")
+            }
+        } else {
+            notify("No files currently marked", "warning")
+        }
+    }
+
+    function deleteAllMarked() {
+        // TODO: Would be nice to have a bulk endpoint for this eventually that takes a string of comma-separated UUIDs
```
Comment (lmarini) [inline app/views/datasets/filesAndFolders.scala.html:139]: Created issue #12

File: app/views/datasets/filesAndFolders.scala.html [hunk 2]
```diff
@@ -27,47 +31,292 @@
...
+            var confirmModal = $(modalHTML);
+            confirmModal.modal();
+            confirmModal.modal("show");
+        } else {
+            notify("No files currently marked", "warning")
+        }
+    }
 
+    function tagMarked() {
+        // TODO: Would be nice to have a bulk endpoint for this eventually that takes a string of comma-separated UUIDs
```
Comment (lmarini) [inline app/views/datasets/filesAndFolders.scala.html:213]: Created issue #13

File: app/views/datasets/filesAndFolders.scala.html [hunk 3]
```diff
@@ -27,47 +31,292 @@
...
+    }
+
+    function clearMarked() {
+        // Set cookie to empty, redraw the file list to get corrected display
+        $.cookie('selected-files-@dataset.id', "");
+        getUpdatedFilesAndFolders();
+    }
+
+    function addFileToMarked(file_id) {
+        // TODO: Use Play framework cookies instead of jquery flavor for better security
```
Comment (lmarini) [inline app/views/datasets/filesAndFolders.scala.html:268]: Created issue #14

File: app/api/Datasets.scala [hunk 4]
```diff
@@ -2427,6 +2440,60 @@ class  Datasets @Inject()(
...
+        // Increment download count for each file in folder
+        folders.get(folderId) match {
+          case Some(fo) => {
+            fo.files.foreach(fid => files.incrementDownloads(fid, user))
+
+            // Use custom enumerator to create the zip file on the fly
+            // Use a 1MB in memory byte array
+            Ok.chunked(enumeratorFromDataset(dataset,1024*1024, -1, bagit, user, None, Some(folderId))).withHeaders(
+              CONTENT_TYPE -> "application/zip",
+              CONTENT_DISPOSITION -> (FileUtils.encodeAttachment(dataset.name+ " ( "+fo.name+" Folder).zip", request.headers.get("user-agent").getOrElse("")))
```
Comment (lmarini) [inline app/api/Datasets.scala:2482]: There is an extra space after the first parenthesis but not before the last. For example: `dataset ( file).zip`

File: app/views/datasets/filesAndFolders.scala.html (hunk 2 above)
Comment (max-zilla) [reply app/views/datasets/filesAndFolders.scala.html]: thanks - removed

File: app/api/Datasets.scala [hunk 5]
```diff
@@ -2427,6 +2440,60 @@ class  Datasets @Inject()(
     }
   }
 
+  def downloadPartial(id: UUID, fileList: String) = PermissionAction(Permission.DownloadFiles, Some(ResourceRef(ResourceRef.dataset, id))) { implicit request =>
```
[the remark this answers is not in the document]
Comment (max-zilla) [reply app/api/Datasets.scala:2448]: added to both

File: app/api/Datasets.scala [hunk 6]
```diff
@@ -2068,7 +2068,9 @@ class  Datasets @Inject()(
     * @return Enumerator to produce array of bytes from a zipped stream containing the bytes of each file
     *         in the dataset
     */
-  def enumeratorFromDataset(dataset: Dataset, chunkSize: Int = 1024 * 8, compression: Int = Deflater.DEFAULT_COMPRESSION, bagit: Boolean, user : Option[User])
+  def enumeratorFromDataset(dataset: Dataset, chunkSize: Int = 1024 * 8,
+                            compression: Int = Deflater.DEFAULT_COMPRESSION, bagit: Boolean,
+                            user : Option[User], fileIDs: Option[List[UUID]], folderId: Option[UUID])
```
[the remark this answers is not in the document]
Comment (max-zilla) [reply app/api/Datasets.scala:2077]: I added more comments. I originally started by splitting into 3 (enumFromDataset, FromFileList, FromFolder) but it was a ton of duplicated code that felt more confusing to look through so I figured 2 new parameters was less jarring.

**Pull request merged**